File: solutions/csharp/annalyns-infiltration/1/AnnalynsInfiltration.cs
static class QuestLogic
{
    public static bool CanFastAttack(bool knightIsAwake)
    {
        bool canAttack = knightIsAwake == true ? false : true;
        return canAttack;
    }

    public static bool CanSpy(bool knightIsAwake, bool archerIsAwake, bool prisonerIsAwake)
    {
        if (knightIsAwake == true || archerIsAwake == true || prisonerIsAwake == true)
        {
            return true;
        }
        else
        {
            return false;
        }
    }

    public static bool CanSignalPrisoner(bool archerIsAwake, bool prisonerIsAwake)
    {
        if (archerIsAwake != true && prisonerIsAwake == true)
        {
            return true;
        }
        return false;
    }

    public static bool CanFreePrisoner(bool knightIsAwake, bool archerIsAwake, bool prisonerIsAwake, bool petDogIsPresent)
    {
        if (petDogIsPresent == true && archerIsAwake != true)
        {
            return true;
        }
        else if (petDogIsPresent != true && prisonerIsAwake == true && knightIsAwake == false && archerIsAwake == false)
        {
            return true;
        }
        else
        {
            return false;
        }
    }
}

File: solutions/go/annalyns-infiltration/1/annalyns_infiltration.go
package annalyn

// CanFastAttack can be executed only when the knight is sleeping.
func CanFastAttack(knightIsAwake bool) bool {
	// const canAttack =  knightIsAwake == true ? false : true
	var canAttack bool
	if knightIsAwake{
		canAttack = false
	} else{
		canAttack = true
	}
	return canAttack
}

// CanSpy can be executed if at least one of the characters is awake.
func CanSpy(knightIsAwake, archerIsAwake, prisonerIsAwake bool) bool {
	var canSpy bool
	if knightIsAwake || archerIsAwake || prisonerIsAwake{
		canSpy = true
	} else{
		canSpy = false
	}
	return canSpy
}

// CanSignalPrisoner can be executed if the prisoner is awake and the archer is sleeping.
func CanSignalPrisoner(archerIsAwake, prisonerIsAwake bool) bool {
	var canSignal bool
	if prisonerIsAwake && !archerIsAwake{
		canSignal = true
	} else{
		canSignal = false
	}
	return canSignal
}

// CanFreePrisoner can be executed if the prisoner is awake and the other 2 characters are asleep
// or if Annalyn's pet dog is with her and the archer is sleeping.
func CanFreePrisoner(knightIsAwake, archerIsAwake, prisonerIsAwake, petDogIsPresent bool) bool {
	var canFree bool
	if (prisonerIsAwake && !knightIsAwake && !archerIsAwake) || (petDogIsPresent && !archerIsAwake){
		canFree = true
	} else{
		canFree = false
	}
	return canFree
}
